src/array.vala

```diff
@@ -92,28 +91,69 @@ public class Array<T>: Object
         return p;
     }
 
-    public unowned T
-    get_scalar(ssize_t [] index)
+    public void*
+    get_pointer ([CCode (array_length = false)] ssize_t[] index)
     {
-        return (T) ((uint8*) _data.get_data () + _offset_for_index (index));
+        return (uint8*) _data.get_data () + _offset_for_index (index);
     }
 
-    public ArrayIterator<T> iterator()
+    public Value
+    get_value ([CCode (array_length = false)] ssize_t[] index)
     {
-        return new ArrayIterator<T>(this);
+        var _value = Value (scalar_type);
+        if (_value.fits_pointer ()) {
+            Memory.copy (_value.peek_pointer (), get_pointer (index), scalar_size);
+        } else if (scalar_type == Type.BOXED) {
+            _value.set_boxed (get_pointer (index));
+        } else if (scalar_type == typeof (string)) {
+            _value.set_string ((string) get_pointer (index));
+        } else if (scalar_type == typeof (double)) {
+            _value.set_double (*((double*) get_pointer (index)));
+        } else {
+            assert_not_reached ();
+        }
+        return _value;
+    }
+
+    public ArrayIterator iterator()
+    {
+        return new ArrayIterator(this);
     }
 
     public void
-    set_scalar(ssize_t [] index, T val)
+    set_pointer ([CCode (array_length = false)] ssize_t[] index, void* val)
     {
         /* What is the best way of doing this the vala way?
          * get triggers a dup function, but looks like there
          * is no clear way doing lvalue?*/
         Memory.copy((uint8*) data.get_data () + _offset_for_index (index), val, scalar_size);
     }
 
+    public void
+    set_value ([CCode (array_length = false)] ssize_t[] index, Value val)
+    {
+        var dest_value = Value (scalar_type);
+        if (val.transform (ref dest_value)) {
+            if (dest_value.fits_pointer ()) {
+                set_pointer (index, dest_value.peek_pointer ());
+            } else if (scalar_type == Type.BOXED) {
+                set_pointer (index, dest_value.get_boxed ());
+            } else if (scalar_type == typeof (string)) {
+                set_pointer (index, val.get_string ());
+            } else if (scalar_type == typeof (double)) {
+                var _ = dest_value.get_double ();
```

Comment:
This is just a start, I need to cover all the type in the union, all other are boxed.